src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,275 @@
+/*-------------------------------------------------------------------------
+ *
+ * intermediate_result_pruning.c
+ *   Functions for pruning intermediate result broadcasting.
+ *
+ * In plain words, each distributed plan pulls the necessary intermediate
+ * results to the worker nodes that the plan hits. This is primarily useful
+ * in three ways. First, if the distributed plan that uses intermediate
+ * result(s) is a router query, then the intermediate results are only
+ * broadcasted to a single node. Second, if a distributed plan consists
+ * of only intermediate results, which is not uncommon, the intermediate
+ * results are broadcasted to a single node only. Finally, if a distributed
+ * query hits a sub-set of the shards in multiple workers, the intermediate
+ * results will be broadcasted to the relevant node(s). The final item becomes
+ * crucial for append/range distributed tables where typically the distributed
+ * queries hit a small subset of shards/workers.
```

Comment:
This is a bit of a wall of text, and will need maintenance, e.g. "Second, if a distributed plan consists of only intermediate results, which is not uncommon, the intermediate results are broadcasted to a single node only." hopefully won't be true for long. 

I think the explanation in the paragraph below is useful, but here maybe just say something like: "We only send intermediate results of subqueries and CTEs to worker nodes that use them in the remainder of the distributed plan to avoid unnecessary network traffic. "